ports/esp32/machine_uart.c

```diff
@@ -146,11 +146,17 @@ STATIC void machine_uart_init_helper(machine_uart_obj_t *self, size_t n_args, co
     mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
     mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
 
-    // wait for all data to be transmitted before changing settings
-    uart_wait_tx_done(self->uart_num, pdMS_TO_TICKS(1000));
+    if (self->uart_num != UART_NUM_0) {  // driver error on repl uart
```

Comment:
what exactly is the "driver error"?  why is it not OK to call `uart_wait_tx_done` on UART 0?

Reply:
```

Reply:
I suspect it's due to the idf UART 0 driver having never been init previously / at startup.

`machine.UART` uses the  IDF "driver/uart.h" interface, wheras repl config in `mphalport.c` bypasses that and goes directly to `uart_tx_one_char()` & `ringbuf_get()` in the underlying "esp32/rom/uart.h" or "esp32c3/rom/uart.h" or "esp32s2/rom/uart.h" etc layer.

Reply:
Interesting... but reconfiguring the baudrate (and changing pins?) is still allowed, even though no driver is installed?

Reply:
Ok, yep I think `uart_wait_tx_done()` is based on monitoring the buffers which are part of the higher layer driver - not used / available in the raw driver.

I've done a quick refactor of esp32/mphalport to use the higher level uart driver, removing the custom interrupt / ringbuffer implementation. With this in place `uart_wait_tx_done()` works fine here.

Reply:
Ok, that might be good, or might not be.  Would need to test REPL performance, eg with `mpremote mount`.

Reply:
> I've done a quick refactor of esp32/mphalport to use the higher level uart driver, removing the custom interrupt / ringbuffer implementation. With this in place uart_wait_tx_done() works fine here.

See #8277.  Sounds like we need to do this.

Reply:
What funny timing! Well my initial attempt is currently broken - not handling raw mode properly at all, presumably an issue with interrupt / control characters but I'm keen to clean it up and do the speed test at least.

Reply:
Also need to test ctrl-C works.  But this should be done as a totally independent commit and PR to this one here, because it's a deep change that needs to be correct.